Test dockerfile in gateway mode

ijc opened the pull request:
This arose as a possibility in https://github.com/moby/buildkit/pull/533#discussion_r205855343. I have a few concerns though:
* This pulls `tonistiigi/dockerfile:latest`, which I suspect (I'm not entirely sure, but I am pretty confident) isn't always going to work e.g. when adding a new dockerfile feature and testing it. Need to find a mechanism to instead inject a locally freshly built image.
* When run independently the `builtin` tests take 3m51 (wallclock) and the `gateway` ones 5m55 (wallclock).
    * a) `gateway` is a surprising amount slower (nb: this was run on my laptop, so not exactly controlled conditions)
    * b) when adding `client` (i.e. the work in #533) as a third mode the overall test breaches the 10min limit imposed by `go test` on a single test (even though it is 3 separate top level `TestFoo` functions). I couldn't see a programmatic way to get around that (could use `go test -timeout` but that's not ideal).
* I marked a bunch of tests as `t.Skip`/TBD in gateway mode since they were failing for me, need to investigate why.

Also:
* Dropped a duplicated `testNoSnapshotLeak`

ijc commented:
Re time taken, perhaps `TestDockerfileMasterGatewayIntegration` should just be gated on a tag? So not run in CI etc but runnable manually?

tonistiigi commented on frontend/dockerfile/dockerfile_test.go:
```diff
@@ -36,48 +36,125 @@ import (
+			if opt.FrontendAttrs == nil {
+				opt.FrontendAttrs = make(map[string]string)
+			}
+			opt.FrontendAttrs["source"] = "tonistiigi/dockerfile"
```
`tonistiigi/dockerfile:master`

ijc replied:
Done, I had misread https://hub.docker.com/r/tonistiigi/dockerfile/tags/. I will also revisit "dockerfile: skip some tests which appear to fail in gateway mode" since this may have changed the results.

ijc replied:
> will also revisit "dockerfile: skip some tests which appear to fail in gateway mode"

All apart from `TestDockerfileInvalidCommand` now pass and I can see the issue with that one so I updated the skip reason and commit message to reflect that.

It does imply that image skew is a real problem though.

tonistiigi commented:
I'd assume the performance is from pulling from registry for every test. We could set up a shared local registry and use that in the tests (and later for busybox/alpine as well).

For the setup I think it might make sense to have a list/map like `allTests` and then if we want to add a new configuration it is enabled with this list. Then we can easily control with env when some configs are disabled and add specific tests that only run on a specific configuration with experimental features. It may also make sense to not run all the workers configuration with the external dockerfile configs by default.

ijc commented:
What about instead of a local registry we produce (once at the start) something which can be used with `buildctl build --import-cache`?

tonistiigi commented:
> You intend this to be a list/map of that matrix of runtime and frontend mode eg. "oci + gateway"/ "containerd + builtin", as opposed to a list of tests like the one in dispatchAll, right?

I'm not 100% sure how to tie it with specific workers but the issue with dispatchAll is that it can only run everything together by only configuring one predefined variable (mode). I think it would be more flexible if the instead of the mode what you would configure would be the list of tests.

ijc commented on frontend/dockerfile/dockerfile_test.go:
```diff
@@ -36,48 +36,122 @@ import (
+	case ModeBuiltin:
+		cmd += " --frontend dockerfile.v0"
+	case ModeMasterGateway:
+		cmd += " --frontend gateway.v0 --frontend-opt=source=tonistiigi/dockerfile"
```
This one should have `:master` on it too. This also explained the failure of `TestDockerfileInvalidCommand` so that commit can now be dropped too.

ijc commented:
@tonistiigi Rereading https://github.com/moby/buildkit/pull/552#issuecomment-410010085 perhaps I overthought it and all you meant was that dispatch all should be:
```go
dispatchAll(t *testing.T, m Mode, []func(*testing.T, integration.Sandbox, *frontend))
```
So that each of `TestDockerfiler*Integeration` could pass a specific list of tests (perhaps utilizing some helper slices of common ones).

tonistiigi commented:
@ijc Builtin is the default on moby so I'm hesitant to drop it atm.

ijc commented:
Ah, I had just assumed it was via the gateway, nm.

ijc commented:
Rebased and added the `Client` variant to the mix, but didn't address any of the other stuff yet since I'm not sure which path to take.

Looks like `Client` mode is leaking some snapshots (`checkAllRemoved` is failing in a bunch of cases). Delving into `testNoSnapshotLeak` it looks like it has leaked an extra snapshot for each of the two local sources. Haven't figured out where/why or what it has to do with client mode though.

ijc commented:
It's a bit of a hack but I added a commit which causes `go test ./frontend/dockerfile` to only run the integration tests in builtin mode by default and to support a `$BUILDKIT_DOCKERFILE_INTEGRATION_MODE` envvar to override, plus added code to `hack/test` to run the other two modes separately if the envvar is not set.

ijc commented:
(nb: this will still fail until #582 is merged)

ijc commented:
I rebased. In theory this should now all pass. The two new tests picked up by the rebase (`testQuotedMetaArgs` and `testIgnoreEntrypoint`) are now both failing the `TestDockerfileMasterGatewayIntegration` case in ways which suggest that the image they are using has not picked up the corresponding fixes -- that's the first bullet in the PR description and I think it's a blocker.

I've dropped the WIP since I think that's the last remaining issue.

ijc commented:
I'm going to see if I can convince `TestDockerfileMasterGatewayIntegration` to use `gateway-devel=true` mode to build the local gateway code ...

ijc commented:
`gateway-devel` mode makes just the `TestDockerfileMasterGatewayIntegration` by itself exceed the 10min barrier.

tonistiigi commented:
@ijc added local mirror support in #618 that should make this much more usable. I'll carry this after it gets merged with some tweaks

ijc commented:
Cool thanks, will watch out for it.

ijc commented:
Closing in favour of #622
